crates/bevy_pbr/src/meshlet/asset.rs

```diff
@@ -68,84 +71,147 @@ pub struct MeshletBoundingSpheres {
-#[derive(Serialize, Deserialize, Copy, Clone, Pod, Zeroable)]
+#[derive(Copy, Clone, Pod, Zeroable)]
 #[repr(C)]
 pub struct MeshletBoundingSphere {
     pub center: Vec3,
     pub radius: f32,
 }
 
 /// An [`AssetLoader`] and [`AssetSaver`] for `.meshlet_mesh` [`MeshletMesh`] assets.
-pub struct MeshletMeshSaverLoad;
+pub struct MeshletMeshSaverLoader;
 
-impl AssetLoader for MeshletMeshSaverLoad {
+impl AssetSaver for MeshletMeshSaverLoader {
     type Asset = MeshletMesh;
     type Settings = ();
+    type OutputLoader = Self;
     type Error = MeshletMeshSaveOrLoadError;
 
-    async fn load<'a>(
+    async fn save<'a>(
         &'a self,
-        reader: &'a mut dyn Reader,
-        _settings: &'a Self::Settings,
-        _load_context: &'a mut LoadContext<'_>,
-    ) -> Result<Self::Asset, Self::Error> {
-        let version = read_u64(reader).await?;
-        if version != MESHLET_MESH_ASSET_VERSION {
-            return Err(MeshletMeshSaveOrLoadError::WrongVersion { found: version });
-        }
-
-        let mut bytes = Vec::new();
-        reader.read_to_end(&mut bytes).await?;
-        let asset = bincode::deserialize_from(FrameDecoder::new(Cursor::new(bytes)))?;
+        writer: &'a mut Writer,
+        asset: SavedAsset<'a, MeshletMesh>,
+        _settings: &'a (),
+    ) -> Result<(), MeshletMeshSaveOrLoadError> {
+        // Write asset version
+        writer
+            .write_all(&MESHLET_MESH_ASSET_VERSION.to_le_bytes())
+            .await?;
 
-        Ok(asset)
-    }
+        // Compress and write asset data
+        writer
+            .write_all(&asset.worst_case_meshlet_triangles.to_le_bytes())
+            .await?;
+        let mut writer = FrameEncoder::new(AsyncWriteSyncAdapter(writer));
+        write_slice(&asset.vertex_data, &mut writer)?;
+        write_slice(&asset.vertex_ids, &mut writer)?;
+        write_slice(&asset.indices, &mut writer)?;
+        write_slice(&asset.meshlets, &mut writer)?;
+        write_slice(&asset.bounding_spheres, &mut writer)?;
+        writer.finish()?;
 
-    fn extensions(&self) -> &[&str] {
-        &["meshlet_mesh"]
+        Ok(())
     }
 }
 
-impl AssetSaver for MeshletMeshSaverLoad {
+impl AssetLoader for MeshletMeshSaverLoader {
     type Asset = MeshletMesh;
     type Settings = ();
-    type OutputLoader = Self;
     type Error = MeshletMeshSaveOrLoadError;
 
-    async fn save<'a>(
+    async fn load<'a>(
         &'a self,
-        writer: &'a mut Writer,
-        asset: SavedAsset<'a, Self::Asset>,
-        _settings: &'a Self::Settings,
-    ) -> Result<(), Self::Error> {
-        writer
-            .write_all(&MESHLET_MESH_ASSET_VERSION.to_le_bytes())
-            .await?;
+        reader: &'a mut dyn Reader,
+        _settings: &'a (),
+        _load_context: &'a mut LoadContext<'_>,
+    ) -> Result<MeshletMesh, MeshletMeshSaveOrLoadError> {
+        // Load and check asset version
+        let version = async_read_u64(reader).await?;
```

Comment:
If you make this read the version number in the native endianness instead of using `from_le_bytes` then you can catch endian mismatches right up front since the version will decode wrong.

Reply:
Confusing error message though, then. I put the endianness check in the plugin itself, since you can't use the asset saver/loader without it. Not that I expect anyone to be using big-endian platforms 😅